packages/basti-cdk/src/basti-instance.ts

```diff
@@ -0,0 +1,183 @@
+   * @default - Public subnets in the VPC
+   */
+  readonly subnetSelection?: aws_ec2.SubnetSelection;
+
+  /**
+   * (Optional) The machine configuration to use for the bastion instance.
+   * If not specified, the default configuration will be used.
+   */
+  readonly machineConfig?: BastiMachineProps;
+
+  /**
+   * (Optional) The tags to apply to the bastion instance. Key-value pairs
+   */
+  readonly tags?: { [key: string]: string };
+}
+
+/**
+ * The basti instance.
+ */
+export class BastiInstance extends Construct {
+  /**
+   * The bastion instance.
+   */
+  public readonly instance: aws_ec2.Instance;
+
+  /**
+   * The bastion instance role.
+   */
+  public readonly role: aws_iam.Role;
+
+  /**
+   * The bastion instance security group.
+   */
+  public readonly securityGroup: aws_ec2.SecurityGroup;
+
+  /**
+   * The ID of the bastion instance.
+   */
+  public readonly bastiId: string;
+
+  /**
+   * The VPC the bastion instance is deployed into.
+   */
+  public readonly vpc: aws_ec2.IVpc;
+
+  /**
+   * Constructs a new instance of the BastiInstance class.
+   *
+   * @param scope CDK construct scope
+   * @param id CDK construct ID
+   * @param props BastiInstanceProps
+   */
+
+  constructor(scope: Construct, id: string, props: BastiInstanceProps) {
+    super(scope, id);
+
+    this.vpc = props.vpc;
+
+    this.bastiId = props.bastiId ?? generateShortId(id);
+
+    const defaultMachineImage = aws_ec2.MachineImage.fromSsmParameter(
+      '/aws/service/ami-amazon-linux-latest/amzn2-ami-kernel-5.10-hvm-x86_64-gp2'
+    );
+
+    const defaultInstanceType = new aws_ec2.InstanceType('t2.micro');
+
+    const defaultSubnetSelection = {
+      subnetType: aws_ec2.SubnetType.PUBLIC,
+    };
+
+    this.role = new aws_iam.Role(this, 'basti-instance-role', {
+      assumedBy: new aws_iam.ServicePrincipal('ec2.amazonaws.com'),
+      roleName: `${BASTION_INSTANCE_ROLE_NAME_PREFIX}-${this.bastiId}`,
+    });
+
+    this.securityGroup = new aws_ec2.SecurityGroup(this, 'basti-instance-sg', {
+      vpc: props.vpc,
+      allowAllOutbound: true,
+      securityGroupName: `${BASTION_INSTANCE_SECURITY_GROUP_NAME_PREFIX}-${this.bastiId}`,
+    });
+
+    this.instance = new aws_ec2.Instance(this, 'basti-instance', {
+      instanceName: `${BASTION_INSTANCE_NAME_PREFIX}-${this.bastiId}`,
+      machineImage: props.machineConfig?.machineImage ?? defaultMachineImage,
+      instanceType: props.machineConfig?.instanceType ?? defaultInstanceType,
+      role: this.role,
+      securityGroup: this.securityGroup,
+      requireImdsv2: true,
+      userData: aws_ec2.UserData.custom(BASTION_INSTANCE_CLOUD_INIT),
+      vpc: props.vpc,
+      vpcSubnets: props.subnetSelection ?? defaultSubnetSelection,
+      ssmSessionPermissions: true,
+    });
+
+    // Add tags to the bastion instance
+    for (const [key, value] of Object.entries(props.tags ?? {})) {
+      // Tags.of(this.instance).add(key, value);
+      Tags.of(this.instance).add(key, value);
```

Comment:
I checked and it does not, should it?

Reply:
I think it should as it might be required by a Tagging Policy ([related comment](https://github.com/BohdanPetryshyn/basti/issues/50#issuecomment-1636734387)) I'm currently working on tags support in `basti init` and I added tagging for all the related resources. Currently, we only create an EBS volume and an ENI.

@DrFunk-n-stein please, correct me if I'm wrong

Reply:
Yup, that's right. 

If tagging is enforced by a combination of tagging policy and service control policy, the creation of basti would simply be denied if the instance is not created with tags applied. 
So adding them afterwards is not an option here unfortunately.

Reply:
See [here](https://aws.amazon.com/blogs/mt/implement-aws-resource-tagging-strategy-using-aws-tag-policies-and-service-control-policies-scps/)

Reply:
@DrFunk-n-stein From all the resources that are created by the construct
* SecurityGroup
* Role
* Instance

Which of those need tags, only the Instance?

As for the instance itself, I'm not entirely sure how to handle this in CDK. The ec2.Instance construct does have the `propagateTagsToVolumeOnCreation` option. How do you normally create these instances, is there an example I can use?

Reply:
Tags are now added to all resources that are created, This might be overkill, but should work.

Reply:
Just to confirm, tags are added to EBS Volumes and ENIs as well?

Reply:
Tags are propagated to the EBS volume. I'm not sure about the ENI, all the things that I could assign tags to now have tags. But I have no idea how CDK/CloudFormation manage these changes. Even if it didn't assign tags to the ENI there is no way (that I know of) to add more tags to the instance.

The `BastiInstance` Construct exposes all resources it creates, so users would be able to add tags in the same way the basti construct can. That should cover all the instances where the Basti construct somehow does not yet assign tags. People that have these more restrictive tagging policies can then contribute their changes back to Basti, and we then integrate them.

Reply:
Also if I understand @DrFunk-n-stein correctly, it is not needed to apply tags to every single resource in CDK, as you can assign Tags to a stack, and they are then propagated to all taggable resources?

Reply:
I think we can remove the `tags` prop altogether according to https://github.com/BohdanPetryshyn/basti/pull/53#discussion_r1270273126

Reply:
I'll rework it, so it is just for the instance then. We have a use case where it is nice just to be able to tag the instance for something. But for the other items it is not relevant. Would that be ok?

Reply:
I think this is doable with the `includeResourceTypes` option of the [Tags.of](https://docs.aws.amazon.com/cdk/v2/guide/tagging.html) method. 

I'd suggest implementing advanced tagging separately after some research on how other popular libraries do that. If we add the property right now without carefully thinking, we'll need to maintain or deprecate it later